### PR TITLE
feat: add 15-budget-intent

### PR DESCRIPTION
**Pull request by anandaroop**
This experiment extracts a structured budget intent from user-supplied free text.

Running this will generate multiple results across a number of examples and models.

I've run a couple of trials and summarized the results here:

https://docs.google.com/spreadsheets/d/1S1-7IJOPc26cgSlWaxP1bz9HEuNJXgQJTDdAyx7XRB0/edit?usp=sharing

Both gpt-4o and claude-haiku seem suitable in terms of accuracy and performance.

Below are some examples of the LLM output

---

### Amounts and ranges

Input:
```
I will spend $500 - $1,000 on this artwork.
```

Output:
```json
{
  "budget": { "currency": "USD", "min": 500, "max": 1000 },
  "numberOfArtworks": 1
}
```

Input:
```
under 1k
```

Output:
```json
{
  "budget": { "currency": "USD", "min": 0, "max": 1000 },
  "numberOfArtworks": 1
}
```

Input:
```
I have $30,000 left in my decorating budget and I’d like to fill the walls with art.
```

Output:
```json
{
  "budget": { "currency": "USD", "min": 0, "max": 30000 },
  "numberOfArtworks": 1
}
```

### Specific timeframe

Input:
```
I’d spend up to $100,000 this year on my collection.
```

Output:
```json
{
  "budget": { "currency": "USD", "min": 0, "max": 100000 },
  "numberOfArtworks": 1,
  "timeSpan": { "number": 1, "unit": "year" }
}
```

### Specific number of artworks

Input:
```
I have 10k to spend on two works
```

Output:
```json
{
  "budget": { "currency": "USD", "min": 0, "max": 10000 },
  "numberOfArtworks": 2
}
```

### Currencies

Input:
```
1000 quid
```

Output:
```json
{
  "budget": { "currency": "GBP", "min": 0, "max": 1000 },
  "numberOfArtworks": 1
}
```

### Unparseable

Input:
```
none of your business
```

Output:
```json
{
  "budget": { "currency": "USD", "min": -1, "max": -1 },
  "numberOfArtworks": 1
}
```

### CODE REVIEW

File: src/15-budget-intent/index.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import { anthropic } from "@ai-sdk/anthropic"
+import { LanguageModel, generateObject } from "ai"
+import { openai } from "@ai-sdk/openai"
+import { z } from "zod"
+import chalk from "chalk"
+import dedent from "dedent"
+import dotenv from "dotenv"
+import fs from "fs"
+import pAll from "p-all"
+import pSeries from "p-series"
```
Comment (anandaroop) [inline src/15-budget-intent/index.ts:26]: In order to fire off 30 async LLM calls in a way that didn't trip any API concurrency limits I had to orchestrate all the `Promise`'d results more carefully.

The promise-handling utils from https://github.com/sindresorhus/promise-fun are invaluable for that.

See L153, L156 below.

File: src/15-budget-intent/index.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+dotenv.config()
+
+const MODELS: Record<string, LanguageModel> = {
+  turbo: openai("gpt-3.5-turbo"),
+  fourO: openai("gpt-4o"),
+  haiku: anthropic("claude-3-haiku-20240307"),
+  // sonnet: anthropic("claude-3-sonnet-20240229"),
+  sonnet35: anthropic("claude-3-5-sonnet-20240620"),
```
Comment (anandaroop) [inline src/15-budget-intent/index.ts:35]: Hot off the presses: https://www.anthropic.com/news/claude-3-5-sonnet

Who can keep up? 😵‍💫

File: src/15-budget-intent/index.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+const system = dedent`
+  Your task is to analyze a plaintext budget statement
+  and produce a structured object to describe it.
+
+  If a budget expresses a range, set min and max accordingly.
+
+  If a budget expresses a single amount, then set min to 0 and max to the amount.
+
+  If a budget cannot be inferred, set both min and max to -1.
```
Comment (anandaroop) [inline src/15-budget-intent/index.ts:66]: I took this expedient route for a budget than can't be inferred. Consumers will basically check `if (intent.budget.max > 0) { doSomething() }` or something like that.

File: src/15-budget-intent/index.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  const exampleThunks = EXAMPLES.map((example) => () => {
+    const modelThunks = _.keys(MODELS).map(
+      (modelName) => () => doLLM(modelName, example)
+    )
+
+    return pAll(modelThunks)
+  })
+
+  const exampleResults = await pSeries(exampleThunks)
+  const flatResults = _.flatten(exampleResults)
```
Comment (anandaroop) [inline src/15-budget-intent/index.ts:157]: Some minor trickiness here re: how to fire off all these requests without getting rate limited — which is what happens when all 30 get fired off immediately via async await.

Instead I wanted to do these in a series of batches.

Had to work my way through the utilities at https://github.com/sindresorhus/promise-fun in order to figure out the best way to do this, but basically I use `pSeries` and `pAll` to :

- fire off the first 5 model requests in parallel for example 1. once all those are received…
- fire off the next 5 model request in parallel for example 2. once all those are received…
- fire off the next 5 model request in parallel for example 3. once all those are received…
- and so on…

File: src/15-budget-intent/index.ts (hunk 1 above)
Comment (mc-jones) [reply src/15-budget-intent/index.ts:26]: 👀 What a great library!

File: src/15-budget-intent/index.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+  fourO: openai("gpt-4o"),
+  haiku: anthropic("claude-3-haiku-20240307"),
+  // sonnet: anthropic("claude-3-sonnet-20240229"),
+  sonnet35: anthropic("claude-3-5-sonnet-20240620"),
+  opus: anthropic("claude-3-opus-20240229"),
+}
+
+const EXAMPLES = [
+  "I’d spend up to $100,000 this year on my collection.",
+  "I have $30,000 left in my decorating budget and I’d like to fill the walls with art.",
```
Comment (mc-jones) [inline src/15-budget-intent/index.ts:41]: I noticed this returned a `1` estimated `numberOfArtworks`. I see why based on the description, but wondering if we might be able to better account for input similar to this?

**Review by mc-jones — APPROVED** [review]
Appreciate the detailed PR 🙏 . Awesome as always.

File: src/15-budget-intent/index.ts (hunk 5 above)
Comment (anandaroop) [reply src/15-budget-intent/index.ts:41]: Good point, I think this is where we might want to recognize an "indefinite number" of artworks as part of the intent, as mentioned [here](https://www.notion.so/artsy/Curated-Discovery-5e43238042014fae853222582a35acd9?pvs=4#459e705629044ccbbbe56d589ad12cf3).

Tinkering with the prompt I am able to produce this behavior, but less reliably with claude-haiku. I guess that's the tiebreaker that tips it to gpt-4o.

Updated in c860557

**Pull request merged**